File: flux-flix-service/src/main/java/com/example/fluxflixservice/MovieController.java
package com.example.fluxflixservice;

import org.springframework.http.MediaType;
import org.springframework.web.bind.annotation.GetMapping;
import org.springframework.web.bind.annotation.PathVariable;
import org.springframework.web.bind.annotation.RequestMapping;
import org.springframework.web.bind.annotation.RestController;
import reactor.core.publisher.Flux;
import reactor.core.publisher.Mono;

/*
    Spring WebMVC-style (reactive) controller
 */
@RestController
@RequestMapping("/movies")
public class MovieController {
    private final MovieService movieService;

    public MovieController(MovieService movieService) {
        this.movieService = movieService;
    }

    @GetMapping
    public Flux<Movie> getMovies() {
        return this.movieService.getAllMovies();
    }

    @GetMapping("/{id}")
    public Mono<Movie> getMovie(@PathVariable String id) {
        return this.movieService.getMovieById(id);
    }

    @GetMapping(value = "/{id}/events", produces = MediaType.TEXT_EVENT_STREAM_VALUE)
    public Flux<MovieEvent> getEvents(@PathVariable String id) {
        return this.movieService.getEventsForMovie(id);
    }
}

File: flux-flix-service-kotlin/src/main/kotlin/com/example/fluxflixservicekotlin/MovieController.kt
package com.example.fluxflixservicekotlin

import org.springframework.http.MediaType
import org.springframework.web.bind.annotation.GetMapping
import org.springframework.web.bind.annotation.PathVariable
import org.springframework.web.bind.annotation.RequestMapping
import org.springframework.web.bind.annotation.RestController
import reactor.core.publisher.Flux
import reactor.core.publisher.Mono

//@RestController
@RequestMapping("/movies")
class MovieController(val movieService: MovieService) {

    @GetMapping
    fun getMovies(): Flux<Movie> = movieService.getAllMovies()

    @GetMapping("/{id}")
    fun getMovie(@PathVariable id: String): Mono<Movie> =
            movieService.getMovieById(id)

    @GetMapping(value = "/{id}/events", produces = arrayOf(MediaType.TEXT_EVENT_STREAM_VALUE))
    fun getEvents(@PathVariable id: String): Flux<MovieEvent> =
            movieService.getEventsForMovie(id)
}